### PR TITLE
Adding crash reporter

### PR DESCRIPTION
**Pull request by royadelparastaran**

### CODE REVIEW

File: crashreporter/src/main/java/com/tealium/crashreporter/CrashReporter.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import com.tealium.internal.listeners.DisableListener
+import com.tealium.internal.listeners.PopulateDispatchListener
+import com.tealium.library.DataSources
+import com.tealium.library.Tealium
+
+/**
+ * Crash Reporter module for tracking crash data. When used, this module populates
+ * <i>crash_*</i> attributes defined in {@link com.tealium.library.DataSources.Key}.
+ */
+class CrashReporter (config: Tealium.Config, tracker: Tracker, truncateStackTrace: Boolean) : Thread.UncaughtExceptionHandler {
```
Comment (tamayok) [inline crashreporter/src/main/java/com/tealium/crashreporter/CrashReporter.kt:15]: CrashReporter should conform to either a Core Collector or Module. Take a look at the other modules for reference. 
`Tealium.Config` does not exist in `tealium-kotlin`. Should be using `TealiumConfig` and can likely be private.
`tracker: Tracker` can be removed, as you won't be needing it. 
Not 100% on this, but I suspect we'll need to incorporate [CoroutineExceptionHandler](https://kotlinlang.org/docs/reference/coroutines/exception-handling.html#exception-handling)

File: crashreporter/src/main/java/com/tealium/crashreporter/Crash.kt [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    const val KEY_STACK_FILENAME = "fileName"
+    const val KEY_STACK_LINE_NUMBER = "lineNumber"
+    const val KEY_STACK_METHOD_NAME = "methodName"
+    const val KEY_THREAD_STATE = "state"
+    const val KEY_THREAD_NUMBER = "threadNumber"
+    const val KEY_THREAD_ID = "threadId"
+    const val KEY_THREAD_PRIORITY = "priority"
+    const val KEY_THREAD_STACK = "stack"
+
+    private val mException: Throwable
```
Comment (tamayok) [inline crashreporter/src/main/java/com/tealium/crashreporter/Crash.kt:13]: `m...` prefixed not used in [Kotlin properties](https://kotlinlang.org/docs/reference/properties.html)

File: crashreporter/src/main/java/com/tealium/crashreporter/Crash.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    private var mThreadNumber: String
+    private var mThreadPriority: String
+
+    init {
+        mException = ex
+        mThread = thread
+    }
+
+
+    fun getCause(): String {
```
Comment (tamayok) [inline crashreporter/src/main/java/com/tealium/crashreporter/Crash.kt:29]: I believe can do without many of these getters. Checkout Kotlin [Getters/Setters](https://kotlinlang.org/docs/reference/properties.html#getters-and-setters)

File: crashreporter/src/main/java/com/tealium/crashreporter/CrashReporter.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import com.tealium.internal.data.Dispatch
+import com.tealium.internal.listeners.ActivityResumeListener
+import com.tealium.internal.listeners.DisableListener
+import com.tealium.internal.listeners.PopulateDispatchListener
+import com.tealium.library.DataSources
+import com.tealium.library.Tealium
+
+/**
+ * Crash Reporter module for tracking crash data. When used, this module populates
+ * <i>crash_*</i> attributes defined in {@link com.tealium.library.DataSources.Key}.
```
Comment (jameskeith) [inline crashreporter/src/main/java/com/tealium/crashreporter/CrashReporter.kt:13]: This is in Javadoc format; kotlin's doc style and the way they create links to other classes is different.
Also, `com.tealium.library.DataSources` doesn't exist as a class in the new Kotlin lib.

File: crashreporter/src/main/java/com/tealium/crashreporter/CrashReporter.kt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ */
+class CrashReporter (config: Tealium.Config, tracker: Tracker, truncateStackTrace: Boolean) : Thread.UncaughtExceptionHandler {
+
+    const val CRASH_COUNT =  "crash_count"
+
+    val mOriginalExceptionHandler: Thread.UncaughtExceptionHandler
```
Comment (jameskeith) [inline crashreporter/src/main/java/com/tealium/crashreporter/CrashReporter.kt:19]: all of these don't need to be public as far as I can see, so would be best to have them `private`

File: crashreporter/src/main/java/com/tealium/crashreporter/CrashReporter.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    val mTracker: Tracker
+    val mTruncateCrashStackTraces: Boolean
+    val mSharedPreferences: SharedPreferences
+    val mCrashCount: Int
+
+    init {
+        mTracker = tracker
+        mTruncateCrashStackTraces = truncateStackTrace
+        mListener = TealiumListener()
+        config.getEventListeners().add(mListener)
```
Comment (jameskeith) [inline crashreporter/src/main/java/com/tealium/crashreporter/CrashReporter.kt:30]: `config.getEventListeners` no longer exists in the kotlin lib. 
You can access them through a `TealiumContext` object which you'd be provided with if you conform your companion object to a ModuleFactory/CollectorFactory.

**Review by jameskeith — COMMENTED** [review]
it's a good start - suggest you update your `build.gradle` to not reference our old Android Java lib:
remove `implementation 'com.tealium:library:5.7.X'` and replace it with `api project(':tealiumlibrary')` that way you'll get compiler warnings for all the classes you no longer have available to use and will need replacing with the kotlin library equivalent.

**Pull request closed without merging**